### PR TITLE
Inherited Extendible

### PR DESCRIPTION
**Pull request by a-canary**
Refactored Extendible class to be used with multiple inheritances, for NetVConn.

* Refactored all access to functional style; ExtGet, ExtSet.
* Added ExtAllocInit, ExtInit, ExtendibleSixeOf to manage lifetime of extendible classes. 
* Added ExtDebugYaml to inspect all values of extendible. 
* Added ExtDebugFormat to render ascii chart of memory alignments of class and extendible memory.

### CODE REVIEW

File: include/tscore/Extendible.h [hunk 1]
```diff
@@ -43,19 +43,21 @@
...
 
+#include "yaml-cpp/yaml.h"
+
 #ifndef ROUNDUP
 #define ROUNDUP(x, y) ((((x) + ((y)-1)) / (y)) * (y))
 #endif
 
 // used for C API
-typedef const void *FieldId_C;
+typedef void *FieldId_C;
```
Comment (SolidWallOfCode) [inline include/tscore/Extendible.h]: `using`?

File: include/tscore/Extendible.h [hunk 2]
```diff
@@ -43,19 +43,21 @@
 #include "tscore/ink_assert.h"
 #include "AcidPtr.h"
 
+#include "yaml-cpp/yaml.h"
+
 #ifndef ROUNDUP
```
Comment (SolidWallOfCode) [inline include/tscore/Extendible.h]: Use an inline function.

File: include/tscore/Extendible.h [hunk 3]
```diff
@@ -150,57 +140,59 @@ template <typename Derived_t> struct Extendible {
     ExtendibleOffset_t offset = 0;                             ///< offset of field from 'this'
     Func_t construct_fn       = nullptr;                       ///< the data type's constructor
     Func_t destruct_fn        = nullptr;                       ///< the data type's destructor
+    std::function<YAML::Node(Extendible const &ext, FieldSchema const &fs)> to_yaml_fn =
```
Comment (SolidWallOfCode) [inline include/tscore/Extendible.h]: Might want to pull the types out as separate `using` types.

**Comment by a-canary** [comment]
Version 9. Simplified functionality of extendible to only be type-safe memory allocator. It no longer attempts to do thread safety as well.

**Comment by a-canary** [comment]
[approve ci]

File: include/tscore/Extendible.h [hunk 4]
```diff
@@ -23,564 +23,861 @@
...
+  Offest_t field_offset;          ///< byte offset from ext_loc to field
+  std::type_index field_type_idx; ///< data type index
+  uint16_t size;                  ///< byte size of field
+  uint8_t align;                  ///< alignment of field
+  uint8_t mask;                   ///< mask for packed bit operations
+
+  // specialize the following
+  std::function<void(FieldPtr)> construct;
+  std::function<void(FieldPtr)> destruct;
+  std::function<YAML::Node(void const *)> toYaml;
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: ok then, how would I create an iterator, where each element is a different type?
Iterating lambdas was the best I could come up with in YAML::convert<Extendible<Derived_t>>

[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: Discussed with Alan. I will replace yaml with ostream, and use a policy template to overload `<<`

File: include/tscore/Extendible.h [hunk 5]
```diff
@@ -23,564 +23,861 @@
...
+
+public:
+  Schema() {}
+  ~Schema();
+
+  /// Testing methods
+  bool no_instances() const; ///< returns true if there are no instances of Extendible<Derived_t>
+  bool reset();              ///< clears all field definitions.
+
+  /// Internal methods
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: This causes a dependency cycle of complex template code. To make these private i  would have to friend 6 templated functions, and forward declare them. And there maybe more use cases for these. I used the '_function' to indicate that it is private. I'll change it to a 'details' namespace.

File: include/tscore/Extendible.h [hunk 6]
```diff
@@ -23,564 +23,861 @@
...
+    Extendible<T> const *e = &t;
+    int ptr_start          = uintptr_t(&(e->ext_loc)) - _base_addr;
+    int ptr_end            = ptr_start + sizeof(Extendible<T>::ext_loc);
+    int ext_start          = uintptr_t(e) + e->ext_loc - _base_addr;
+    int ext_end            = Extendible<T>::schema._fullSize(ext_start);
+
+    ink_assert(ptr_end <= ext_start);
+    ink_assert(ext_end <= _full_size);
+
+    printf("\n%*s | EXT  | %5ib |", 30, typeid(T).name(), ext_end - ext_start);
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: I'll rewrite to just return a string.

File: include/tscore/Extendible.h [hunk 7]
```diff
@@ -23,564 +23,861 @@
...
+// UTILITY Functions
 
-  /// C API - returns pointer, no internal thread safety
-  void *get(FieldId_C &field);
+//////////////////////////////////////////////////////
+/// HexString functions for untyped C storage
 
-  // operator[]
-  template <AccessEnum Access_t, typename Field_t> auto operator[](FieldId<Access_t, Field_t> field) { return get(field); }
+std::string
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: This is the only thing that would depend on BufferWriter, for 3 lines of code. So I'll hold off on this.

File: include/tscore/Extendible.h [hunk 8]
```diff
@@ -23,564 +23,861 @@
...
+};
 
-    AccessEnum access         = NUM_ACCESS_TYPES;              ///< which API is used to access the data
-    std::type_index type      = std::type_index(typeid(void)); ///< datatype
-    ExtendibleOffset_t size   = 0;                             ///< size of field
-    ExtendibleOffset_t offset = 0;                             ///< offset of field from 'this'
-    Func_t construct_fn       = nullptr;                       ///< the data type's constructor
-    Func_t destruct_fn        = nullptr;                       ///< the data type's destructor
+bool
+bufFromHexString(void const *buf, uint16_t size, std::string const &hex_str)
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: removed.

File: include/tscore/Extendible.h [hunk 9]
```diff
@@ -23,564 +23,861 @@
...
-  static auto destruct_fn  = [](void *ptr) { static_cast<Field_t *>(ptr)->~Field_t(); };
+  EXT::Offest_t const *loc = (EXT::Offest_t const *)(uintptr_t(derived) + desc.ext_loc_offset);
+  return FieldPtr(uintptr_t(derived) + (*loc) + desc.field_offset);
+}
 
-  fields[field_name]  = FieldSchema{Access_t, std::type_index(typeid(Field_t)), size, 0, construct_fn, destruct_fn};
-  field_id.offset_ptr = &fields[field_name].offset;
-  updateMemOffsets();
-  return true;
+namespace EXT
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: changed.

File: include/tscore/Extendible.h [hunk 10]
```diff
@@ -23,564 +23,861 @@
...
+template <typename Derived_t, typename Field_t>
+Field_t const &
+ExtFieldGet(void const *fld_ptr, ExtFieldId<Derived_t, Field_t> const &field)
+{
+  return *static_cast<Field_t const *>(fld_ptr);
+}
+
+template <typename Derived_t, typename Field_t>
+Field_t &
+ExtFieldSet(void *fld_ptr, ExtFieldId<Derived_t, Field_t> const &field)
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: So I have the public API prefixed as `::ExtClassName`, and the internal API as `ext::ClassName`. This just happened organically, is there a better practice to use?

File: include/tscore/Extendible.h [hunk 11]
```diff
@@ -23,564 +23,861 @@
...
 
-  uint32_t acc_offset = ROUNDUP(sizeof(Derived_t), alloc_align);
+////////////////////////////////////////////////////
+// Schema Methods
 
-  ExtendibleOffset_t size_blocks[] = {4, 2, 1};
+Schema::~Schema() {}
+
+void
+Schema::_updateMemOffsets()
```
[the remark this answers is not in the document]
Comment (a-canary) [reply include/tscore/Extendible.h]: moved to Extendible.cc

**Comment by a-canary** [comment]
rebased and squashed after making changes based on feedback

**Comment by a-canary** [comment]
[approve ci]

**Comment by a-canary** [comment]
[approve ci]

**Comment by a-canary** [comment]
[approve ci autest]

**Comment by a-canary** [comment]
Resolved rebased and merge conflicts.

**Pull request merged**